src/MobileRedirectManager.php

```diff
@@ -0,0 +1,225 @@
+		return isset( $_COOKIE[ self::DISABLED_COOKIE_NAME ] ) && '1' === $_COOKIE[ self::DISABLED_COOKIE_NAME ];
+	}
+
+	/**
+	 * Output the mobile redirection Javascript code.
+	 */
+	public static function add_mobile_redirect_script() {
+		?>
+		<script>
+			(function ( ampSlug, disabledCookieName, userAgents ) {
+				const currentUserAgent = navigator.userAgent;
+				const isMobile = userAgents.some( ( userAgent ) => currentUserAgent.includes( userAgent ) );
+
+				if ( isMobile ) {
+					document.addEventListener( 'DOMContentLoaded', () => {
+						// Show the mobile version switcher link once the DOM has loaded.
+						const siteVersionSwitcher = document.getElementById( 'site-version-switcher' );
+						if ( siteVersionSwitcher ) {
+							siteVersionSwitcher.hidden = false;
+						}
+
+						// Disable mobile redirection if the user opts for the mobile version.
+						const versionSwitchLink = document.getElementById( 'version-switch-link' );
+						if ( versionSwitchLink ) {
+							versionSwitchLink.addEventListener( 'click', () => {
+								document.cookie = `${disabledCookieName}=;path=/;samesite=strict${ 'https:' === location.protocol ? ';secure' : '' }`;
+							} )
+						}
+					} );
+				}
+
+				const mobileRedirectionDisabled = document.cookie
+						.split(';')
+						.some( ( item ) => `${ disabledCookieName }=1` === item.trim() );
+
+				// Short-circuit if mobile redirection is disabled.
+				if ( mobileRedirectionDisabled ) {
+					return;
+				}
+
+				const url = new URL( location.href );
+				if ( isMobile && ! url.searchParams.has( ampSlug ) ) {
+					window.stop(); // Stop loading the page! This should cancel all loading resources.
+
+					// Replace the current page with the AMP version.
+					url.searchParams.append( ampSlug, '1' );
+					location.replace( url.href );
+				}
+			} )(
+				<?php echo wp_json_encode( amp_get_slug() ); ?>,
+				<?php echo wp_json_encode( self::DISABLED_COOKIE_NAME ); ?>,
+				<?php echo wp_json_encode( self::get_user_agents() ); ?>
+			)
+		</script>
+		<?php
+	}
+
+	/**
+	 * Output the markup for the mobile version switcher.
+	 *
+	 * @param bool   $is_amp Modifies markup to be AMP compatible if true.
+	 * @param string $url    URL to canonical version of page.
+	 * @param string $text   Text for the anchor element.
+	 */
+	public static function add_mobile_version_switcher_markup( $is_amp, $url, $text ) {
+		$rel         = $is_amp ? 'amphtml' : 'noamphtml';
+		$hidden_attr = $is_amp ? '' : 'hidden';
+		?>
+		<style>
+			#version-switch-link {
+				display: block;
+				width: 100%;
+				padding: 15px 0;
+				font-size: 16px;
+				font-weight: 600;
+				color: #eaeaea;
+				text-align: center;
+				background-color: #444;
+				border: 0;
+			}
+		</style>
+		<div id="site-version-switcher" <?php echo esc_attr( $hidden_attr ); ?>>
+			<a
+				id="version-switch-link"
+				rel="<?php echo esc_attr( $rel ); ?>"
+				href="<?php echo esc_url( $url ); ?>"
```

Comment:
As discussed, if `$is_amp` then this needs to have an `onclick` attribute which clears the cookie so that mobile redirection will be resumed.

Reply:
Done in 831726e399559c1b3fd17d023f23d95cdc2f8d50.